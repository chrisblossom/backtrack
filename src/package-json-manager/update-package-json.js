/* @flow */

import { cloneDeep, isEmpty, isPlainObject, get, merge, unset } from 'lodash';
import { getParentsFromPath, mapObjectKeyNames } from '../utils/object-utils';
import type { PackageJson } from '../types.js';

function updatePackageJson(
    packageJson: PackageJson,
    managedKeys: ?PackageJson,
    previousManagedKeys: ?PackageJson,
): PackageJson {
    /**
     * merge mutates. Deep clone to prevent mutation
     */
    const packageJsonCopy = cloneDeep(packageJson);

    /**
     * remove no longer managed keys
     */
    if (previousManagedKeys) {
        const mapPreviousManagedKeys = mapObjectKeyNames(previousManagedKeys);

        mapPreviousManagedKeys.forEach((key) => {
            const pathTree = getParentsFromPath(key);

            pathTree.forEach((currentPath, index) => {
                /**
                 * the first path is the missing managed key
                 */
                if (index === 0) {
                    unset(packageJsonCopy, currentPath);
                }

                /**
                 * Unset object empty tree
                 */
                const matchedPackageKey = get(packageJsonCopy, currentPath);
                if (
                    isPlainObject(matchedPackageKey) &&
                    isEmpty(matchedPackageKey)
                ) {
                    unset(packageJsonCopy, currentPath);
                }
            });
        });
    }

    const result = merge(packageJsonCopy, managedKeys);

    return result;
}

export { updatePackageJson };
