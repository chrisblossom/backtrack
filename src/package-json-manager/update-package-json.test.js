/* @flow */

require('./update-package-json');

const updatePackageJson = (packageJson, managedKeys, previousManagedKeys) =>
    require('./update-package-json').updatePackageJson(
        // $FlowIgnore
        packageJson,
        managedKeys,
        // $FlowIgnore
        previousManagedKeys,
    );

describe('updatePackageJson', () => {
    test('merges keys', () => {
        const packageJson = {
            name: 'package',
            scripts: {
                one: 'script',
            },
            dependencies: {
                lodash: '4',
            },
            devDependencies: {},
            peerDependencies: {},
            optionalDependencies: {},
        };

        const managedKeys = {
            scripts: {
                dev: 'backtrack dev',
                test: 'jest',
            },
            dependencies: {
                underscore: '4',
            },
        };

        const result = updatePackageJson(packageJson, managedKeys);

        expect(result).toMatchSnapshot();
    });

    test('removes keys no longer managed', () => {
        const packageJson = {
            scripts: { one: 'script', dev: 'backtrack dev' },
        };

        const managedKeys = { scripts: { dev: 'backtrack dev' } };

        const previouslyManagedKeys = {
            scripts: { one: 'script', dev: 'backtrack dev' },
        };

        const result = updatePackageJson(
            packageJson,
            managedKeys,
            previouslyManagedKeys,
        );

        expect(result).toEqual({ scripts: { dev: 'backtrack dev' } });
    });

    test('removes keys nested root no longer managed', () => {
        const packageJson = {
            name: 'backtrack',
            testing: {
                nested: 'one',
            },
        };

        const managedKeys = {};

        const previouslyManagedKeys = {
            testing: {
                nested: 'one',
            },
        };

        const result = updatePackageJson(
            packageJson,
            managedKeys,
            previouslyManagedKeys,
        );

        expect(result).toEqual({ name: 'backtrack' });
    });

    test('removes item from array when item removed', () => {
        const packageJson = {
            name: 'backtrack',
            files: ['one', 'two'],
        };

        const managedKeys = {
            files: ['one'],
        };

        const previouslyManagedKeys = {
            files: ['one', 'two'],
        };

        const result = updatePackageJson(
            packageJson,
            managedKeys,
            previouslyManagedKeys,
        );

        expect(result).toEqual({ name: 'backtrack', files: ['one'] });
    });

    test('removes keys with .', () => {
        const packageJson = {
            name: 'backtrack',
            scripts: {
                'lint.fix': 'one',
                nested: {
                    'test.ed': {
                        remove: 'this',
                    },
                },
            },
            're.move': 'this',
        };

        const managedKeys = {};

        const previouslyManagedKeys = {
            scripts: {
                'lint.fix': 'one',
                nested: {
                    'test.ed': {
                        remove: 'this',
                    },
                },
            },
            're.move': 'this',
        };

        const result = updatePackageJson(
            packageJson,
            managedKeys,
            previouslyManagedKeys,
        );

        expect(result).toEqual({ name: 'backtrack' });
    });

    test('removes all empty even deeper but stops when not empty', () => {
        const packageJson = {
            random: 'key',
            testing: {
                nested: {
                    other_random: 'key',
                    nested2: {
                        nested3: 'all',
                    },
                },
            },
        };

        const managedKeys = {
            other: 'hello',
        };

        const previouslyManagedKeys = {
            testing: {
                nested: {
                    nested2: {
                        nested3: 'all',
                    },
                },
            },
        };

        const result = updatePackageJson(
            packageJson,
            managedKeys,
            previouslyManagedKeys,
        );

        expect(result).toEqual({
            random: 'key',
            other: 'hello',
            testing: {
                nested: {
                    other_random: 'key',
                },
            },
        });
    });

    test('works if already removed', () => {
        const packageJson = { scripts: { dev: 'backtrack dev' } };
        const managedKeys = { scripts: { dev: 'backtrack dev' } };
        const previouslyManagedKeys = {
            scripts: { one: 'script', dev: 'backtrack dev' },
        };

        const result = updatePackageJson(
            packageJson,
            managedKeys,
            previouslyManagedKeys,
        );

        expect(result).toEqual({ scripts: { dev: 'backtrack dev' } });
    });
});
