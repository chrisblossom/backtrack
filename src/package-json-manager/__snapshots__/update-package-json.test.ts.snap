// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updatePackageJson merges keys 1`] = `
{
  "dependencies": {
    "lodash": "4",
    "underscore": "4",
  },
  "devDependencies": {},
  "name": "package",
  "optionalDependencies": {},
  "peerDependencies": {},
  "scripts": {
    "dev": "backtrack dev",
    "one": "script",
    "test": "jest",
  },
}
`;
